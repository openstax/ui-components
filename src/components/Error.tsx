import React from "react";
import { ModalBody, ModalBodyHeading } from "./Modal";
import styled from "styled-components/macro";
import { colors } from "../../src/theme";
import { ErrorContext } from "../contexts";

export interface ErrorPropTypes {
  className?: string;
  children?: React.ReactNode;
  heading?: string;
}

const EventId = styled.div`
  font-size: 0.9rem;
  color: ${colors.palette.neutralMedium};
  margin-top: 1.2rem;
`;

export const Error = ({ heading, children, ...props }: ErrorPropTypes) => {
  const context = React.useContext(ErrorContext);

  return <ModalBody {...props} data-testid='error'>
    <ModalBodyHeading>{heading ?? `Uh-oh, there's been a glitch`}</ModalBodyHeading>
    {children ?? <>
      We're not quite sure what went wrong. Restart your browser. If this doesn't solve
      the problem, visit our <a href="https://openstax.secure.force.com/help" target="_blank">Support Center</a>.
    </>}
    <EventId>{context?.eventId}</EventId>
  </ModalBody>
};
