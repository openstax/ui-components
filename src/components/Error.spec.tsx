import renderer from 'react-test-renderer';
import { render, act } from '@testing-library/react';
import { Error } from './Error';
import * as Sentry from '@sentry/react';

describe('Error', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(
      <Error />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can override text', () => {
    const tree = renderer.create(
      <Error heading='An important heading'>Body text</Error>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('updates lastEventId when Sentry.lastEventId() changes', async () => {
    jest.useFakeTimers();

    const setIntervalSpy = jest.spyOn(global, 'setInterval');
    const sentrySpy = jest.spyOn(Sentry, 'lastEventId');
    sentrySpy.mockReturnValue(undefined);

    const { getByTestId } = render(<Error />);

    const errorElement = getByTestId('event-id');
    expect(errorElement.textContent).toBe('');
    expect(setIntervalSpy).toHaveBeenCalled();

    act(() => {
      sentrySpy.mockReturnValue('updatedEventId');
      jest.advanceTimersByTime(100);
    });

    expect(errorElement.textContent).toContain('updatedEventId');

    jest.restoreAllMocks();
  });
});
