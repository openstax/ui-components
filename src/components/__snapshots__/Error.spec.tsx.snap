// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error can override text 1`] = `
<div
  className="sc-dkzDqf iJAkzW"
  data-testid="error"
>
  <h3
    className="sc-gsnTZi jDaiGk"
  >
    An important heading
  </h3>
  Body text
  <div
    className="sc-hKMtZM bkppVE"
    data-testid="event-id"
  />
</div>
`;

exports[`Error matches snapshot 1`] = `
<div
  className="sc-dkzDqf iJAkzW"
  data-testid="error"
>
  <h3
    className="sc-gsnTZi jDaiGk"
  >
    Uh-oh, there's been a glitch
  </h3>
  We're not quite sure what went wrong. Restart your browser. If this doesn't solve the problem, visit our 
  <a
    href="https://openstax.secure.force.com/help"
    target="_blank"
  >
    Support Center
  </a>
  .
  <div
    className="sc-hKMtZM bkppVE"
    data-testid="event-id"
  />
</div>
`;
